File: problems/Max Area of Island/maxAreaOfIsland.cs
public class Solution {
    public int MaxAreaOfIsland(int[][] grid) {
        int m = grid.Length, n = grid[0].Length;
        var visitedSet = new HashSet<(int, int)>();
        int maxArea = 0;

        for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            if (grid[i][j] == 1 && !visitedSet.Contains((i, j)))
                maxArea = Math.Max(maxArea, Dfs(grid, m, n, i, j, visitedSet));

        return maxArea;
    }

    int Dfs(int[][] grid, int m, int n, int row, int col, HashSet<(int, int)> visitedSet, int count = 1)
    {
        if (row >= m || row < 0 || col < 0 || col >= n || grid[row][col] == 0 ||
            visitedSet.Contains((row, col)))
            return 0;
        visitedSet.Add((row, col));
        count += Dfs(grid, m, n, row + 1, col, visitedSet);
        count += Dfs(grid, m, n, row - 1, col, visitedSet);
        count += Dfs(grid, m, n, row, col - 1, visitedSet);
        count += Dfs(grid, m, n, row, col + 1, visitedSet);

        return count;
    }
}

File: problems/Max Area of Island/maxAreaOfIsland.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var maxAreaOfIsland = function(grid) {
    const rows = grid.length;
    const cols = grid[0].length;
    let maxSize = 0;
    let sizeArray = [];

    const dfs = function(g, r, c) {
        if (0 > r || g.length <= r || 0 > c || g[0].length <= c || 0 === g[r][c]) {
            return;
        }

        g[r][c] = 0;
        ++sizeArray[sizeArray.length - 1];

        dfs(g, r - 1, c);
        dfs(g, 1 + r, c);
        dfs(g, r, c - 1);
        dfs(g, r, 1 + c);
    };

    for (let row = 0; rows > row; ++row) {
        for (let col = 0; cols > col; ++col) {
            if (1 === grid[row][col]) {
                sizeArray.push(0);
                dfs(grid, row, col);
                maxSize = Math.max(maxSize, sizeArray[sizeArray.length - 1]);
            }
        }
    }

    return maxSize;
};
